### ISSUE TITLE
Another "Could not get class storage for"... issue

### ISSUE THREAD

**Issue by Synchro** [vimeo/psalm#3247]
I've found several issues containing this "Could not get class storage for...". Most of them have a fix that seems to work for their authors, and another suggested using `--clear-cache` and `--no-cache`. I'm running Psalm 3.11.2 on PHP 7.4.5 and I'm getting the same problem leading to a stack trace like this:

```
./vendor/bin/psalm --no-cache 
Scanning files...
Analyzing files...

EUncaught Exception: Could not get class storage for myclass
Stack trace in the forked worker:
#0 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/ClassLikes.php(1511): Psalm\Internal\Provider\ClassLikeStorageProvider->get('myclass')
#1 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Type/Reconciler.php(480): Psalm\Internal\Codebase\ClassLikes->getConstantForClass('myclass', '$conf', 4, NULL)
#2 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Type/Reconciler.php(236): Psalm\Type\Reconciler::getValueForKey(Object(Psalm\Codebase), 'myclass::...', Array, Array, Object(Psalm\CodeLocation), true, false, false, false, false)
#3 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOpAnalyzer.php(187): Psalm\Type\Reconciler::reconcileKeyedTypes(Array, Array, Array, Array, Array, Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Array, false, Object(Psalm\CodeLocation))
#4 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php(337): Psalm\Internal\Analyzer\Statements\Expression\BinaryOpAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\BinaryOp\BooleanAnd), Object(Psalm\Context), 0, false)
#5 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfAnalyzer.php(649): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\BinaryOp\BooleanAnd), Object(Psalm\Context))
#6 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfAnalyzer.php(90): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyzeIfConditional(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\BinaryOp\BooleanAnd), Object(Psalm\Context), Object(Psalm\Codebase), Object(Psalm\Internal\Scope\IfScope), 52168)
#7 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(374): Psalm\Internal\Analyzer\Statements\Block\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))
#8 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php(541): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), Object(Psalm\Context), true)
#9 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(702): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Internal\Provider\NodeDataProvider), Object(Psalm\Context))
#10 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(190): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)
#11 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/IncludeAnalyzer.php(158): Psalm\Internal\Analyzer\FileAnalyzer->analyze(Object(Psalm\Context), false, NULL)
#12 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php(743): Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), NULL)
#13 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(736): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\Include_), Object(Psalm\Context), false, NULL, true)
#14 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(190): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)
#15 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(357): Psalm\Internal\Analyzer\FileAnalyzer->analyze(NULL)
#16 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(184): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(0, '/Users/marcus/S...')
#17 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(423): Psalm\Internal\Fork\Pool->__construct(Array, Object(Closure), Object(Closure), Object(Closure), Object(Closure))
#18 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(267): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 23)
#19 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(568): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 23, false, true)
#20 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/psalm.php(588): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/marcus/S...', false)
#21 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/psalm(2): require_once('/Users/marcus/S...')
#22 {main} in /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php:348
Stack trace:
#0 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(380): Psalm\Internal\Fork\Pool->readResultsFromChildren()
#1 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(484): Psalm\Internal\Fork\Pool->wait()
#2 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(267): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 23)
#3 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(568): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 23, false, true)
#4 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/psalm.php(588): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/marcus/S...', false)
#5 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/psalm(2): require_once('/Users/marcus/S...')
#6 {main}
(Psalm 3.11.2@d470903722cfcbc1cd04744c5491d3e6d13ec3d9 crashed due to an uncaught Throwable)
```

As other issues mentioned, the class I'm expecting it to look for is named `MyClass`, not `myclass`, though that class does indeed have a property called `$conf`.

I don't know what psalm is trying to do at this point. The code it's analysing is a legacy project, so it is likely to have issues for psalm to find, but it all runs fine.

Another problem is that the stack trace truncates output at critical points, for example:

    #4 /Users/marcus/Sites/myproject/vendor/vimeo/psalm/src/psalm.php(588): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/marcus/S...', false)

and that means I can't see what file it's looking at when it crashes which makes this hard to investigate. I can't see a php setting to fix that - `log_errors_max_len` doesn't change it.

I tried with `--debug-by-line` as well, and with that the output stops at a specific line, which is the first line of this `if`:

    if (array_key_exists('cookiedomain', MyClass::$conf['system']) &&
        !empty(MyClass::$conf['system']['cookiedomain'])
    ) {

This line is in a file which is excluded in psalm's config, but is loaded by files that are analysed by psalm.

I also noticed that after this dump, psalm produces no more output, but doesn't actually quit - I have to ctrl-C to make it return to a prompt.

Any idea what I'm doing wrong?

**Comment by Synchro** [comment]
If I paste in just the function that line occurs in, psalm reports no issues. If I paste the whole file (it's big!), I get:

    Psalm runner output: 

    Internal Psalm error on line undefined - undefined

**Comment by muglug** [comment]
@Synchro the best you could do here would be to try to reproduce the issue in a separate GitHub repo, paring the project down to the bare essentials.

**Comment by muglug** [comment]
Any reason you closed this?

**Comment by Synchro** [comment]
Unfortunately it's not public code so I can't share it. It's really not that critical at the moment; I'm getting useful results out of phpcs & phpstan to help spot issues in this project. I don't have time now to understand what's going wrong in psalm, and I don't want to clog up your issues with unreproducible problems.

**Comment by robinkooli** [comment]
Experiencing this also.

**Comment by tdtm** [comment]
I was encountering that and after a bit of digging, I found that this was the cause:

```php
/**
 * @var FooController
 * @psalm-scope-this Fully\Qualified\Class\Name\FooController
 */
$this->doSomething();

// ...etc
```
...without any import for `FooController`. I fixed it by importing `FooController`, but it also worked if I made ` * @var FooController` fully qualified.

I suppose the crash was caused by trying to load `FooController` without the fully qualified reference for the `@var` annotation.
